File: leetCode/easy/1365-countNumsSmallerThanCurrNum.ts
import { assertEquals } from "https://deno.land/std@0.182.0/testing/asserts.ts";
import { describe, it } from "https://deno.land/std@0.182.0/testing/bdd.ts";

function countNumsSmallerThanCurrNum(nums: number[]) {
  return nums.reduce((acc: number[], outerNum) => {
    const result = nums.reduce(
      (result, innerNum) => (innerNum < outerNum ? (result += 1) : result),
      0
    );
    acc.push(result);

    return acc;
  }, []);
}

export { countNumsSmallerThanCurrNum };

describe("countNumsSmallerThanCurrNum", () => {
  it("1. should return an array of the number of smaller numbers to the right of each index", () => {
    const nums = [8, 1, 2, 2, 3];
    assertEquals(countNumsSmallerThanCurrNum(nums), [4, 0, 1, 1, 3]);
  });

  it("2. should return an array of the number of smaller numbers to the right of each index", () => {
    const nums = [6, 5, 4, 8];
    assertEquals(countNumsSmallerThanCurrNum(nums), [2, 1, 0, 3]);
  });

  it("3. should return an array of the number of smaller numbers to the right of each index", () => {
    const nums = [7, 7, 7, 7];
    assertEquals(countNumsSmallerThanCurrNum(nums), [0, 0, 0, 0]);
  });
});

File: leetCode/easy/1365-countNumsSmallerThanCurrNum.jl
function countNumsSmallerThanCurrNum(nums::Vector{Int})
  result::Vector{Int} = []

  for (i, iVal) in enumerate(nums)
    count = 0
    for (j, jVal) in enumerate(nums)
      if jVal < iVal
        count += 1
      end
    end

    push!(result, count)
  end

  return result
end

export countNumsSmallerThanCurrNum

using Test

Test.@testset "countNumsSmallerThanCurrNum" begin
  Test.@test countNumsSmallerThanCurrNum([8, 1, 2, 2, 3]) == [4, 0, 1, 1, 3]
  Test.@test countNumsSmallerThanCurrNum([6, 5, 4, 8]) == [2, 1, 0, 3]
  Test.@test countNumsSmallerThanCurrNum([7, 7, 7, 7]) == [0, 0, 0, 0]
end